### PR TITLE
D/w 7434792 b

### PR DESCRIPTION
**Pull request by jfeingold35**

### CODE REVIEW

**Comment by michaelhoefer** [comment]
@jfeingold35 I'm a little surprised the command requires the original location and language in order to remove rules.  How again are they expected to remember those they used previously?

File: pmd-cataloger/src/main/java/sfdc/sfdx/scanner/pmd/catalog/PmdCatalogCategory.java [hunk 1]
```diff
@@ -9,14 +9,20 @@
 	 * The path to the definition file for this category.
 	 */
 	private String path;
+	/**
+	 * The JAR containing this category's definition file.
+	 */
+	private String sourceJar;
```
Comment (michaelhoefer) [inline pmd-cataloger/src/main/java/sfdc/sfdx/scanner/pmd/catalog/PmdCatalogCategory.java:15]: This is not required, right?  They don't have to have specified a jar file originally?

Comment (jfeingold35) [reply pmd-cataloger/src/main/java/sfdc/sfdx/scanner/pmd/catalog/PmdCatalogCategory.java:15]: Currently, PMD rules have to be defined in a JAR. We're not at the point where we could just accept the root directory of a Java project.

**Comment by jfeingold35** [comment]
@michaelhoefer , how else would the user be able to specify the rules they want deleted if not by their location? I do agree that demanding they specify the language might be a bit excessive, but I was thinking about the case where the user has a folder that contains a large number of JARs that define rules for many different languages, and they want to delete only the Apex-related JARs in that folder. Then they can do `sfdx scanner:rule:remove --language apex --path 'my/folder'`. I suppose we could make `--language` optional, with the default behavior being to remove matching JARs for all languages.
As far as how the user is expected to remember the location, I think the solution to that would be to add some way for them to log out that information. Like if part of our `list` or `describe` command was to include the JAR where the rule is defined.

**Comment by michaelhoefer** [comment]
I think we do need to make language optional.  it makes sense to me that way, yes.  And yes, I think we need a way of listing what is available to be removed.  I don't want to overcomplicate this, but even just letting them run scanner:rule:remove with no args and having it list the available paths to nuke would be great.

**Comment by jfeingold35** [comment]
The "no args == show options" idea is great. I'll implement that.

File: src/lib/CustomRulePathManager.ts [hunk 2]
```diff
@@ -94,6 +94,54 @@ export class CustomRulePathManager extends AsyncCreatable {
...
+		// Expand the patterns into actual paths. E.g., expand directories into the rule objects they contain, etc.
+		const expandedPaths = await this.expandPaths(paths);
+
+		// Now that we've got the possible paths, we need to see which ones are actually present.
+		return expandedPaths.filter((p) => {
+			// Determine the engine associated with this path.
+			const e = this.determineEngineForPath(p);
+			// If there's nothing mapped for that engine, or the engine has nothing for this language, we can drop this
+			// path.
+			if (!this.pathsByLanguageByEngine.has(e) || !this.pathsByLanguageByEngine.get(e).has(language)) {
```
Comment (rmohan20) [inline src/lib/CustomRulePathManager.ts:111]: Thinking out loud. Highly possible that the user adds a language in lowercase and tries to delete passing in the same language in mixed case. As of today, we do this only in the Java layer. We should rather be unifying the case before storing in CustomPath.json.

Comment (jfeingold35) [reply src/lib/CustomRulePathManager.ts:111]: You're proposing always casting the language argument to lowercase? That makes sense to me.

File: messages/remove.js [hunk 3]
```diff
@@ -0,0 +1,27 @@
+module.exports = {
+	"commandDescription": "[Short description]",
```
Comment (rmohan20) [inline messages/remove.js:2]: Can we add the messages right away? We wouldn't have to remember to revisit later, especially now that our release cycle is kind of spontaneous.

Comment (jfeingold35) [reply messages/remove.js:2]: Yeah, I forgot to do that. I'll get on it ASAP.

**Comment by jfeingold35** [comment]
@michaelhoefer , @rmohan20 , seeking your opinion: What should be the behavior when `--language` is provided but `--path` isn't? The options are:
- Delete all paths for the provided language, as an inverse of the way that providing just a path would delete that path for all languages.
- Show only the available paths for that language, like a constricted result of the output when you add no parameters.
- Just disallow that configuration.
What do you think? I'm leaning towards Option 1, because that feels pretty intuitive to me, but maybe you guys disagree.

**Comment by rmohan20** [comment]
> @michaelhoefer , @rmohan20 , seeking your opinion: What should be the behavior when `--language` is provided but `--path` isn't? The options are:
> 
>     * Delete all paths for the provided language, as an inverse of the way that providing just a path would delete that path for all languages.
> 
>     * Show only the available paths for that language, like a constricted result of the output when you add no parameters.
> 
>     * Just disallow that configuration.
>       What do you think? I'm leaning towards Option 1, because that feels pretty intuitive to me, but maybe you guys disagree.

I agree, Option1 does seem more intuitive. We can display the paths deleted as part of output.

**Comment by michaelhoefer** [comment]
I'm not so sure.  It is intuitive, yes, but I'm not convinced it is a valid use case.  I would not want you to spend too much time on something that won't be used in practice.  So while I agree that option 1 is the most  correct, option 3 may be the most appropriate.  Depends on the work and testing involved.

**Pull request merged**